File: VB/WindowsApplication1/CustomEditor/MyButtonEdit.vb
Imports System
Imports System.ComponentModel
Imports System.Drawing
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Controls
Imports DevExpress.Utils.Drawing
Imports DevExpress.XtraEditors.ViewInfo

Namespace WindowsApplication1

    Public Class MyButtonEdit
        Inherits ButtonEdit
        Implements IAutoHeightControl

        Shared Sub New()
            RepositoryItemMyButtonEdit.Register()
        End Sub

        Public Overrides ReadOnly Property EditorTypeName As String
            Get
                Return RepositoryItemMyButtonEdit.EditorName
            End Get
        End Property

        <DesignerSerializationVisibility(DesignerSerializationVisibility.Content)>
        Public Overloads ReadOnly Property Properties As RepositoryItemMyButtonEdit
            Get
                Return TryCast(MyBase.Properties, RepositoryItemMyButtonEdit)
            End Get
        End Property

        Private ReadOnly Property SupportsAutoHeight As Boolean Implements IAutoHeightControl.SupportsAutoHeight
            Get
                Return True
            End Get
        End Property

        Public Event eHeightChanged As EventHandler

        Private Custom Event HeightChanged As EventHandler Implements IAutoHeightControl.HeightChanged
            AddHandler(ByVal value As EventHandler)
                AddHandler eHeightChanged, value
            End AddHandler

            RemoveHandler(ByVal value As EventHandler)
                RemoveHandler eHeightChanged, value
            End RemoveHandler

            RaiseEvent(ByVal sender As Object, ByVal e As EventArgs)
                RaiseEvent eHeightChanged(sender, e)
            End RaiseEvent
        End Event

        Protected Sub RaiseHeightChanged()
            RaiseEvent eHeightChanged(Me, EventArgs.Empty)
        End Sub

        Private Function CalcHeight(ByVal cache As GraphicsCache) As Integer Implements IAutoHeightControl.CalcHeight
            If ViewInfo.IsReady Then
                Dim ih As IHeightAdaptable = TryCast(ViewInfo, IHeightAdaptable)
                If ih IsNot Nothing Then Return ih.CalcHeight(cache, Width)
            End If

            Return Height
        End Function
    End Class
End Namespace

File: CS/WindowsApplication1/CustomEditor/MyButtonEdit.cs
using System;
using System.ComponentModel;
using System.Drawing;
using System.Windows.Forms;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraEditors.Registrator;
using DevExpress.XtraEditors.Controls;
using DevExpress.Utils.Drawing;
using DevExpress.XtraEditors.ViewInfo;

namespace WindowsApplication1
{

    public class MyButtonEdit : ButtonEdit, IAutoHeightControl
    {
        static MyButtonEdit()
        {
            RepositoryItemMyButtonEdit.Register();
        }
     
        public override string EditorTypeName
        {
            get { return RepositoryItemMyButtonEdit.EditorName; }
        }
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
        public new RepositoryItemMyButtonEdit Properties
        {
            get { return base.Properties as RepositoryItemMyButtonEdit; }
        }


        bool IAutoHeightControl.SupportsAutoHeight { get { return true; } }
        public event EventHandler heightChanged;
        event EventHandler IAutoHeightControl.HeightChanged
        {
            add { heightChanged += value; }
            remove { heightChanged -= value; }
        }
        protected void RaiseHeightChanged()
        {
            if (heightChanged != null)
                heightChanged(this, EventArgs.Empty);
        }

        int IAutoHeightControl.CalcHeight(GraphicsCache cache)
        {
            if (ViewInfo.IsReady)
            {
                IHeightAdaptable ih = ViewInfo as IHeightAdaptable;
                if (ih != null) return ih.CalcHeight(cache, Width);
            }
            return Height;

        }
    }
}
